client/core/core.go

```diff
@@ -1586,6 +1586,32 @@ func (c *Core) WalletSettings(assetID uint32) (map[string]string, error) {
 	return dbWallet.Settings, nil
 }
 
+func (c *Core) setAppPass(pw []byte) error {
+	if len(pw) == 0 {
+		return fmt.Errorf("empty password not allowed")
+	}
+
+	crypter := c.newCrypter(pw)
+	err := c.db.Store(keyParamsKey, crypter.Serialize())
+	if err != nil {
+		return fmt.Errorf("error storing key parameters: %w", err)
+	}
+
+	return nil
+}
+
+// ChangeAppPass updates the application password to the provided new password
+// after validating the current password.
+func (c *Core) ChangeAppPass(appPW, newAppPW []byte) error {
+	// Validate current password.
+	_, err := c.encryptionKey(appPW)
+	if err != nil {
+		return newError(authErr, "ChangePassApp password error: %v", err)
+	}
+	// Set new password.
+	return c.setAppPass(newAppPW)
```

Comment:
Things that are encrypted with the old pw:
- each `xcWallet`'s pass (for protected wallets) in `encPW`/`pw` (see `(*xcWallet).Unlock`), in memory and DB
- each `dexAccount`'s private key in `encKey`/`privKey` (see `(*dexAccount).setupEncryption`), in memory and DB

These "enc" fields need to be updated and the DB entries patched.

I believe your suggestion in matrix that a `db.UpdateAccount` or perhaps a more purposeful `UpdateAccountKey` method would be needed for this.  Fortunately there's already a `db.SetWalletPassword` that re-saves the `encPW`.

Reply:
With https://github.com/decred/dcrdex/pull/978/commits/2a725e20a2ed3f0b03bd54b83ec962fb8354ba73
I'm able to change app password lock & unlock wallets but when trying to place an order I get:
```
2021-03-05 02:06:17.245 [ERR] WEB: error placing order: new order request with DEX server 127.0.0.1:17273 market dcr_btc failed: rpc error: error code 33: account ID mismatch

```

server's side:
```
2021-03-05 02:05:15.294 [INF] AUTH: New user registered: acct cc934689ce3a489a9f6cb19551d44c4351b852ad6f47cd5fddedf2c6fb9d45fd from 127.0.0.1 paid 100000000 to SsbeD1czwoAwqxV8fiq1sbhT7JsypYN5LPK
2021-03-05 02:05:15.394 [TRC] AUTH: User cc934689ce3a489a9f6cb19551d44c4351b852ad6f47cd5fddedf2c6fb9d45fd cancellation rate is now NaN% (0 cancels : 0 successes). Violation = false
2021-03-05 02:05:15.439 [DBG] AUTH: User cc934689ce3a489a9f6cb19551d44c4351b852ad6f47cd5fddedf2c6fb9d45fd score = 0: 0 (violations) + 0 (0 preimage misses) - 0 (0 successes)
2021-03-05 02:05:15.456 [INF] AUTH: Authenticated account cc934689ce3a489a9f6cb19551d44c4351b852ad6f47cd5fddedf2c6fb9d45fd from 127.0.0.1 with 0 active orders, 0 active matches
2021-03-05 02:06:17.245 [DBG] AUTH: Handling of 'limit' request for user cc934689ce3a489a9f6cb19551d44c4351b852ad6f47cd5fddedf2c6fb9d45fd failed: error code 33: account ID mismatch


```

Reply:
Oh gosh, seeing this diff you pushed I think I may have misguided you.  I might have been seeing way more work than necessary.  The password encrypts the _encryption keys themselves_ in the DB... and I actually don't think there is any change to the encryption parameters themselves (the serialized Crypter), and thus all the encPW and encKey fields might even stay the same.

:grimacing:

I gotta open up the branch and look at it.

Referring to this:

```go
func (c *Core) encryptionKey(pw []byte) (encrypt.Crypter, error) {
	keyParams, err := c.db.Get(keyParamsKey)
	if err != nil {
		return nil, fmt.Errorf("key retrieval error: %v", err)
	}
	crypter, err := c.reCrypter(pw, keyParams)
	if err != nil {
		return nil, fmt.Errorf("encryption key deserialization error: %v", err)
	}
	return crypter, nil
}
```

The app password decreypts (via reCrypter) the keyParams from the DB to get the crypter.  And (crap) I think that the serialized crypter should be the same after a successful password change, with only the encrypted key data (keyParams) being changed, which would mean all the enc fields aren't any different either.  

I'm going to look at this in the next couple hours and help straighten it out, sorry.

Reply:
Nah, false alarm.  Changing the password changes he encryption key as it's derived from the password (not encrypted by the password).
I'll look into what's causing that account id issue.